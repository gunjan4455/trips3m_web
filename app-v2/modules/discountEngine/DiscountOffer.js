import React from 'react';
import PropTypes from 'prop-types';
import { SubmissionError } from 'redux-form';

import './discountOffer.scss';
import { OffersBoxIcon } from 'helpers/Icon/Icon';
import { trackSegment } from 'actions/segmentEvents';
import { ORGANISM_CLICKED, ORGANISM_VIEWED } from 'actions/segmentEvents';
import Modal from 'modules/shared/GenericModal';
import DiscountOfferPopup from './components/DiscountOfferPopup';
import { OffersIcon } from 'helpers/Icon/Icon';
import AppliedCouponBanner from "./components/AppliedCouponBanner";

export default class DiscountOffer extends React.Component {

  static propTypes = {
    tripId: PropTypes.number.isRequired,
    offers: PropTypes.array,
    renderAppliedCoupon: PropTypes.bool,
    appliedOffer: PropTypes.object,
    params: PropTypes.object.isRequired,
    discountInformation: PropTypes.object.isRequired,
    applyDiscountCoupon: PropTypes.func.isRequired,
    removeDiscountCoupon: PropTypes.func.isRequired,
    updateDiscountCoupon: PropTypes.func.isRequired,
    updateFormField: PropTypes.func.isRequired,
    clearFormSubmitError: PropTypes.func.isRequired,
    fetchAppliedOffer: PropTypes.func.isRequired
  };

  static defaultProps = {
    offers: [],
    renderAppliedCoupon: false,
    appliedOffer: {}
  };

  constructor(props) {
    super(props);
    this.state = {
      closeModal: false
    };

    this.hasDiscountApplied = props.discountInformation && props.discountInformation.discount > 0 || false;
    this.isEventTracked = false;
  }

  componentDidMount() {
    if (this.props.discountInformation.discountableTypeId) {
      this.props.fetchAppliedOffer(this.props.discountInformation.discountableTypeId);
    }
  }

  UNSAFE_componentWillReceiveProps(nextProps) {
    if (this.props.discountInformation.discountableTypeId !== nextProps.discountInformation.discountableTypeId) {
      this.props.fetchAppliedOffer(nextProps.discountInformation.discountableTypeId);
    }
  }

  trackSegmentEvent = (data, event) => {
    if (event === ORGANISM_VIEWED && this.isEventTracked) {
      return;
    }
    this.isEventTracked = event === ORGANISM_VIEWED;

    const { tripId } = this.props;

    trackSegment({
      category: data.category,
      section: data.section,
      object: data.object,
      cta: data.cta,
      event: event,
      trip_id: tripId,
      destination_id: '',
      misc_id_1: data.miscId1,
      misc_type_1: data.miscType1,
      misc_id_2: data.miscId2,
      misc_type_2: data.miscType2,
      misc_id_3: data.miscId3,
      misc_type_3: data.miscType3,
      misc_id_4: data.miscId4,
      misc_type_4: data.miscType4,
      misc_id_5: data.miscId5,
      misc_type_5: data.miscType5
    });
  };

  throwSubmissionError = (error) => {
    throw new SubmissionError({
      _error: error
    });
  };

  handleRemoveDiscountCoupon = () => {
    return this.props.removeDiscountCoupon(this.props.tripId).then((results) => {
      if (results && results.valid === 'true') {
        window.location.reload();
      }
    });
  };

  applyCouponCallback = (results) => {
    if (results) {
      if (results.valid === 'true') {
        this.setState({ closeModal: true });
        window.location.reload();
      }
      else {
        this.throwSubmissionError(results.msg);
      }
    }
    else {
      this.throwSubmissionError('Some server error occurred!');
    }
  };

  applyDiscountCoupon = (params) => {
    return this.props.applyDiscountCoupon(params)
      .then(this.applyCouponCallback);
  };

  handleSubmit = (values) => {
    this.props.clearFormSubmitError();

    if (!values.discountOfferInput && (!values.discountOfferRadio || values.discountOfferRadio === '0')) {
      this.throwSubmissionError('Please select atleast 1 coupon');
    }

    const params = { offerType: 'coupon', offerCode: '', tripId: this.props.tripId };
    let applyMode = 'Typed';
    let couponId = null;

    if (values.discountOfferInput) {
      params.offerCode = values.discountOfferInput;
    }
    else if (values.discountOfferRadio) {
      const offer = this.props.offers.filter(offer => offer.couponType.toLowerCase() === 'normal'
        && offer.couponId.toString() === values.discountOfferRadio)[0];
      params.offerCode = offer.code;
      couponId = offer.couponId;
      applyMode = 'Selection';
    }

    this.trackSegmentEvent({
      category: '',
      section: 'Discount Pop-up',
      object: '',
      cta: 'Apply',
      miscId1: couponId,
      miscType1: 'Coupon Applied',
      miscId2: applyMode,
      miscType2: 'Application Mode',
    }, ORGANISM_CLICKED);

    if (this.hasDiscountApplied) {
      return this.props.updateDiscountCoupon(params)
        .then(this.applyCouponCallback);
    }

    return this.applyDiscountCoupon(params);
  };

  onSensorChange = (isVisible) => {
    if (isVisible) {
      this.trackSegmentEvent({
        category: '',
        section: 'Discount Banner',
        object: '',
        cta: '',
      }, ORGANISM_VIEWED);
    }
  };

  renderDiscountOffersModal() {
    const {
      offers, updateFormField, discountInformation
    } = this.props;

    return (<Modal
      trigger={<button type="button" className="sfc1 f14 fw7 m0 ml5 flex alignCenter cursorP at_getOffersCTA"><span
        className="input-24 mr8 p2 iblock"><OffersIcon /></span> Get Offers</button>}
      fullView={true}
      isDefaultModal={false}
      closeModal={this.state.closeModal}
      onAfterModalOpen={() => this.trackSegmentEvent({ cta: 'Change', section: 'Discount Banner' }, ORGANISM_CLICKED)}
      onAfterModalClose={() => this.trackSegmentEvent({ cta: 'Close', section: 'Discount Pop-up' }, ORGANISM_CLICKED)}
      hasCustomClose={true}
    >
      <DiscountOfferPopup
        offers={offers}
        discountInformation={discountInformation}
        handleSubmit={this.handleSubmit}
        updateFormField={updateFormField}
        trackSegmentEvent={this.trackSegmentEvent} />
    </Modal>);
  }

  render() {
    return (
      this.props.renderAppliedCoupon
        ? <AppliedCouponBanner appliedOffer={this.props.appliedOffer}
                               removeDiscountCoupon={this.handleRemoveDiscountCoupon} />
        : this.renderDiscountOffersModal()
    );
  }
}
