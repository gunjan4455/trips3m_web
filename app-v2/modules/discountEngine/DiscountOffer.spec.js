import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, configure } from 'enzyme';

import DiscountOffer from './DiscountOffer';

configure({ adapter: new Adapter() });
let component;

const discountInfoData = {
  tripId: '3918699',
  offers: [
    {
      couponId: 393,
      type: 'Coupon',
      code: '1BB2D9D87',
      heading: 'TEST_NE_Jan_1000_Referral',
      desc: 'Rs. 1000 Off ',
      tnc: '- A flat discount value of Rs. 1000 will be applied to the quotes.\r\n- The discounts will be applicable only after Trip is converted.\r\n- If the Trip is cancelled after booking, discount value will be reverted.',
      offerVal: 1000,
      couponType: 'Normal',
      endDate: '2019-07-21',
      discountType: 1,
      discountCurrency: 'Rupee'
    },
    {
      couponId: 395,
      type: 'Coupon',
      code: '1BB2D9D86',
      heading: 'TEST2_NE_Jan_1000_Referral',
      desc: 'Rs. 1000 Off ',
      tnc: '- A flat discount value of Rs. 1000 will be applied to the quotes.\r\n- The discounts will be applicable only after Trip is converted.\r\n- If the Trip is cancelled after booking, discount value will be reverted.',
      offerVal: 1000,
      couponType: 'Normal',
      endDate: '2019-07-21',
      discountType: 1,
      discountCurrency: 'Rupee'
    },
    {
      couponId: 394,
      type: 'Coupon',
      code: '1BB2D9D88',
      heading: 'TEST1_NE_Jan_1000_Referral',
      desc: 'Rs. 1000 Off ',
      tnc: '- A flat discount value of Rs. 1000 will be applied to the quotes.\r\n- The discounts will be applicable only after Trip is converted.\r\n- If the Trip is cancelled after booking, discount value will be reverted.',
      offerVal: 1000,
      couponType: 'View Only',
      endDate: '2019-07-21',
      discountType: 2,
      discountCurrency: 'Rupee'
    }
  ],
  appliedOffer: {},
  params: {
    quoteid: "7124217",
    requestedTripId: "3918699"
  },
  discountInformation: {
    discount: 0,
    discountableTypeId: null,
    discountableType: null
  },
  applyDiscountCoupon: jest.fn,
  removeDiscountCoupon: jest.fn,
  updateDiscountCoupon: jest.fn,
  updateFormField: jest.fn,
  clearFormSubmitError: jest.fn,
};

const createDiscountOfferComponent = () => {
  return (<DiscountOffer {...discountInfoData} />);
};

describe('Discount Engine: Render Correctly ', () => {

  beforeAll(() => {
    const DiscountOfferComponent = createDiscountOfferComponent();
    component = shallow(DiscountOfferComponent);
  });

  // it('should have correct error message', () => {
  //   const { mountedComponent } = component;
  //   console.log('mountedComponent......', JSON.stringify(mountedComponent));
  //   mountedComponent.find('.at_discountOfferFormSubmit').simulate('click');
  //   expect(mountedComponent.find('.input-error-text').text().trim().toLowerCase()).toEqual('Please select atleast 1 coupon');
  // });
  // it('should show offers list on get offer click', () => {
  //   component.find('Get Offers').simulate('click');
  //   expect(component.find('Current Discounts & Offers').toBeTruthy());
  // });

  it('should match with the previous snapshot', () => {
    expect(component).toMatchSnapshot();
  });
});
