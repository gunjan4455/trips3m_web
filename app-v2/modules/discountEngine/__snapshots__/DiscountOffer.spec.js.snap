// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Discount Engine: Render Correctly  should match with the previous snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <DiscountOffer
    appliedOffer={Object {}}
    applyDiscountCoupon={[Function]}
    clearFormSubmitError={[Function]}
    discountInformation={
      Object {
        "discount": 0,
        "discountableType": null,
        "discountableTypeId": null,
      }
    }
    offers={
      Array [
        Object {
          "code": "1BB2D9D87",
          "couponId": 393,
          "couponType": "Normal",
          "desc": "Rs. 1000 Off ",
          "discountCurrency": "Rupee",
          "discountType": 1,
          "endDate": "2019-07-21",
          "heading": "TEST_NE_Jan_1000_Referral",
          "offerVal": 1000,
          "tnc": "- A flat discount value of Rs. 1000 will be applied to the quotes.
- The discounts will be applicable only after Trip is converted.
- If the Trip is cancelled after booking, discount value will be reverted.",
          "type": "Coupon",
        },
        Object {
          "code": "1BB2D9D86",
          "couponId": 395,
          "couponType": "Normal",
          "desc": "Rs. 1000 Off ",
          "discountCurrency": "Rupee",
          "discountType": 1,
          "endDate": "2019-07-21",
          "heading": "TEST2_NE_Jan_1000_Referral",
          "offerVal": 1000,
          "tnc": "- A flat discount value of Rs. 1000 will be applied to the quotes.
- The discounts will be applicable only after Trip is converted.
- If the Trip is cancelled after booking, discount value will be reverted.",
          "type": "Coupon",
        },
        Object {
          "code": "1BB2D9D88",
          "couponId": 394,
          "couponType": "View Only",
          "desc": "Rs. 1000 Off ",
          "discountCurrency": "Rupee",
          "discountType": 2,
          "endDate": "2019-07-21",
          "heading": "TEST1_NE_Jan_1000_Referral",
          "offerVal": 1000,
          "tnc": "- A flat discount value of Rs. 1000 will be applied to the quotes.
- The discounts will be applicable only after Trip is converted.
- If the Trip is cancelled after booking, discount value will be reverted.",
          "type": "Coupon",
        },
      ]
    }
    params={
      Object {
        "quoteid": "7124217",
        "requestedTripId": "3918699",
      }
    }
    removeDiscountCoupon={[Function]}
    renderAppliedCoupon={false}
    tripId="3918699"
    updateDiscountCoupon={[Function]}
    updateFormField={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <DiscountOfferPopup
        appliedOfferId={0}
        discountInformation={
          Object {
            "discount": 0,
            "discountableType": null,
            "discountableTypeId": null,
          }
        }
        handleSubmit={[Function]}
        hasDiscountApplied={false}
        offers={
          Array [
            Object {
              "code": "1BB2D9D87",
              "couponId": 393,
              "couponType": "Normal",
              "desc": "Rs. 1000 Off ",
              "discountCurrency": "Rupee",
              "discountType": 1,
              "endDate": "2019-07-21",
              "heading": "TEST_NE_Jan_1000_Referral",
              "offerVal": 1000,
              "tnc": "- A flat discount value of Rs. 1000 will be applied to the quotes.
- The discounts will be applicable only after Trip is converted.
- If the Trip is cancelled after booking, discount value will be reverted.",
              "type": "Coupon",
            },
            Object {
              "code": "1BB2D9D86",
              "couponId": 395,
              "couponType": "Normal",
              "desc": "Rs. 1000 Off ",
              "discountCurrency": "Rupee",
              "discountType": 1,
              "endDate": "2019-07-21",
              "heading": "TEST2_NE_Jan_1000_Referral",
              "offerVal": 1000,
              "tnc": "- A flat discount value of Rs. 1000 will be applied to the quotes.
- The discounts will be applicable only after Trip is converted.
- If the Trip is cancelled after booking, discount value will be reverted.",
              "type": "Coupon",
            },
            Object {
              "code": "1BB2D9D88",
              "couponId": 394,
              "couponType": "View Only",
              "desc": "Rs. 1000 Off ",
              "discountCurrency": "Rupee",
              "discountType": 2,
              "endDate": "2019-07-21",
              "heading": "TEST1_NE_Jan_1000_Referral",
              "offerVal": 1000,
              "tnc": "- A flat discount value of Rs. 1000 will be applied to the quotes.
- The discounts will be applicable only after Trip is converted.
- If the Trip is cancelled after booking, discount value will be reverted.",
              "type": "Coupon",
            },
          ]
        }
        trackSegmentEvent={[Function]}
        updateFormField={[Function]}
      />,
      "closeModal": false,
      "customClass": "",
      "customCloseTrigger": false,
      "disabled": false,
      "fullView": true,
      "hasCustomClose": true,
      "hasNoClose": false,
      "isDefaultModal": false,
      "isTriggerFullView": false,
      "isWhiteIcon": false,
      "onAfterModalClose": [Function],
      "onAfterModalOpen": [Function],
      "onRequestCloseCustom": [Function],
      "trigger": <button
        className="sfc1 f14 fw7 m0 ml5 flex alignCenter cursorP at_getOffersCTA"
        type="button"
      >
        <span
          className="input-24 mr8 p2 iblock"
        >
          <IconSvg />
        </span>
         Get Offers
      </button>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "appliedOfferId": 0,
        "discountInformation": Object {
          "discount": 0,
          "discountableType": null,
          "discountableTypeId": null,
        },
        "handleSubmit": [Function],
        "hasDiscountApplied": false,
        "offers": Array [
          Object {
            "code": "1BB2D9D87",
            "couponId": 393,
            "couponType": "Normal",
            "desc": "Rs. 1000 Off ",
            "discountCurrency": "Rupee",
            "discountType": 1,
            "endDate": "2019-07-21",
            "heading": "TEST_NE_Jan_1000_Referral",
            "offerVal": 1000,
            "tnc": "- A flat discount value of Rs. 1000 will be applied to the quotes.
- The discounts will be applicable only after Trip is converted.
- If the Trip is cancelled after booking, discount value will be reverted.",
            "type": "Coupon",
          },
          Object {
            "code": "1BB2D9D86",
            "couponId": 395,
            "couponType": "Normal",
            "desc": "Rs. 1000 Off ",
            "discountCurrency": "Rupee",
            "discountType": 1,
            "endDate": "2019-07-21",
            "heading": "TEST2_NE_Jan_1000_Referral",
            "offerVal": 1000,
            "tnc": "- A flat discount value of Rs. 1000 will be applied to the quotes.
- The discounts will be applicable only after Trip is converted.
- If the Trip is cancelled after booking, discount value will be reverted.",
            "type": "Coupon",
          },
          Object {
            "code": "1BB2D9D88",
            "couponId": 394,
            "couponType": "View Only",
            "desc": "Rs. 1000 Off ",
            "discountCurrency": "Rupee",
            "discountType": 2,
            "endDate": "2019-07-21",
            "heading": "TEST1_NE_Jan_1000_Referral",
            "offerVal": 1000,
            "tnc": "- A flat discount value of Rs. 1000 will be applied to the quotes.
- The discounts will be applicable only after Trip is converted.
- If the Trip is cancelled after booking, discount value will be reverted.",
            "type": "Coupon",
          },
        ],
        "trackSegmentEvent": [Function],
        "updateFormField": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <DiscountOfferPopup
          appliedOfferId={0}
          discountInformation={
            Object {
              "discount": 0,
              "discountableType": null,
              "discountableTypeId": null,
            }
          }
          handleSubmit={[Function]}
          hasDiscountApplied={false}
          offers={
            Array [
              Object {
                "code": "1BB2D9D87",
                "couponId": 393,
                "couponType": "Normal",
                "desc": "Rs. 1000 Off ",
                "discountCurrency": "Rupee",
                "discountType": 1,
                "endDate": "2019-07-21",
                "heading": "TEST_NE_Jan_1000_Referral",
                "offerVal": 1000,
                "tnc": "- A flat discount value of Rs. 1000 will be applied to the quotes.
- The discounts will be applicable only after Trip is converted.
- If the Trip is cancelled after booking, discount value will be reverted.",
                "type": "Coupon",
              },
              Object {
                "code": "1BB2D9D86",
                "couponId": 395,
                "couponType": "Normal",
                "desc": "Rs. 1000 Off ",
                "discountCurrency": "Rupee",
                "discountType": 1,
                "endDate": "2019-07-21",
                "heading": "TEST2_NE_Jan_1000_Referral",
                "offerVal": 1000,
                "tnc": "- A flat discount value of Rs. 1000 will be applied to the quotes.
- The discounts will be applicable only after Trip is converted.
- If the Trip is cancelled after booking, discount value will be reverted.",
                "type": "Coupon",
              },
              Object {
                "code": "1BB2D9D88",
                "couponId": 394,
                "couponType": "View Only",
                "desc": "Rs. 1000 Off ",
                "discountCurrency": "Rupee",
                "discountType": 2,
                "endDate": "2019-07-21",
                "heading": "TEST1_NE_Jan_1000_Referral",
                "offerVal": 1000,
                "tnc": "- A flat discount value of Rs. 1000 will be applied to the quotes.
- The discounts will be applicable only after Trip is converted.
- If the Trip is cancelled after booking, discount value will be reverted.",
                "type": "Coupon",
              },
            ]
          }
          trackSegmentEvent={[Function]}
          updateFormField={[Function]}
        />,
        "closeModal": false,
        "customClass": "",
        "customCloseTrigger": false,
        "disabled": false,
        "fullView": true,
        "hasCustomClose": true,
        "hasNoClose": false,
        "isDefaultModal": false,
        "isTriggerFullView": false,
        "isWhiteIcon": false,
        "onAfterModalClose": [Function],
        "onAfterModalOpen": [Function],
        "onRequestCloseCustom": [Function],
        "trigger": <button
          className="sfc1 f14 fw7 m0 ml5 flex alignCenter cursorP at_getOffersCTA"
          type="button"
        >
          <span
            className="input-24 mr8 p2 iblock"
          >
            <IconSvg />
          </span>
           Get Offers
        </button>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "appliedOfferId": 0,
          "discountInformation": Object {
            "discount": 0,
            "discountableType": null,
            "discountableTypeId": null,
          },
          "handleSubmit": [Function],
          "hasDiscountApplied": false,
          "offers": Array [
            Object {
              "code": "1BB2D9D87",
              "couponId": 393,
              "couponType": "Normal",
              "desc": "Rs. 1000 Off ",
              "discountCurrency": "Rupee",
              "discountType": 1,
              "endDate": "2019-07-21",
              "heading": "TEST_NE_Jan_1000_Referral",
              "offerVal": 1000,
              "tnc": "- A flat discount value of Rs. 1000 will be applied to the quotes.
- The discounts will be applicable only after Trip is converted.
- If the Trip is cancelled after booking, discount value will be reverted.",
              "type": "Coupon",
            },
            Object {
              "code": "1BB2D9D86",
              "couponId": 395,
              "couponType": "Normal",
              "desc": "Rs. 1000 Off ",
              "discountCurrency": "Rupee",
              "discountType": 1,
              "endDate": "2019-07-21",
              "heading": "TEST2_NE_Jan_1000_Referral",
              "offerVal": 1000,
              "tnc": "- A flat discount value of Rs. 1000 will be applied to the quotes.
- The discounts will be applicable only after Trip is converted.
- If the Trip is cancelled after booking, discount value will be reverted.",
              "type": "Coupon",
            },
            Object {
              "code": "1BB2D9D88",
              "couponId": 394,
              "couponType": "View Only",
              "desc": "Rs. 1000 Off ",
              "discountCurrency": "Rupee",
              "discountType": 2,
              "endDate": "2019-07-21",
              "heading": "TEST1_NE_Jan_1000_Referral",
              "offerVal": 1000,
              "tnc": "- A flat discount value of Rs. 1000 will be applied to the quotes.
- The discounts will be applicable only after Trip is converted.
- If the Trip is cancelled after booking, discount value will be reverted.",
              "type": "Coupon",
            },
          ],
          "trackSegmentEvent": [Function],
          "updateFormField": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
